---
title: "Differentiating_between_DA_Neuron_subtypes"
output: html_document
---

In this markdown we attempt to differentiate between the two different DA neuron subtypes we identified as clusters 12 and 18. 

First we show that these clusters are highly likely to be DA neurons by showing they express DA specific markers


```{R}
library(Seurat)
library(cowplot)
library(ggplot2)
library(ggsci)
library(dplyr)
library(psych)
library(pheatmap)
library(harmony)
#library(clusterProfiler)
library(org.Hs.eg.db)
library(DOSE)
library(GOSemSim)
library(enrichplot)

```

```{R}
AllMB <- readRDS("Files/")



```

```{R}
VlnPlot(AllMB, features = c("TH"),pt.size = 0) +
        theme(axis.text = element_text(size=8),
              axis.title = element_text(size = 12),
              legend.text = element_text(size = 8),
              title = element_text(size = 12),
              legend.key.size = unit(0.4,"cm"))

THVln <- VlnPlot(AllMB, features = c("TH"),pt.size = 0) +
        theme(axis.text = element_text(size=8),
              axis.title = element_text(size = 12),
              legend.text = element_text(size = 8),
              title = element_text(size = 12),
              legend.key.size = unit(0.4,"cm"))

ggsave(THVln, filename = "Files/TH_Vln_scRNA_seq_Batch4.pdf", device = "pdf", width = 6, height = 4, units = "in")
```
SLC6A3, 

```{R}
VlnPlot(AllMB, features = c("SLC6A3"),pt.size = 0) +
        theme(axis.text = element_text(size=8),
              axis.title = element_text(size = 12),
              legend.text = element_text(size = 8),
              title = element_text(size = 12),
              legend.key.size = unit(0.4,"cm"))

SLC6A3Vln <- VlnPlot(AllMB, features = c("SLC6A3"),pt.size = 0) +
        theme(axis.text = element_text(size=8),
              axis.title = element_text(size = 12),
              legend.text = element_text(size = 8),
              title = element_text(size = 12),
              legend.key.size = unit(0.4,"cm"))

ggsave(SLC6A3Vln, filename = "Files/SLC6A3_Vln_scRNA_seq_Batch4.pdf", device = "pdf", width = 6, height = 4, units = "in")
```


```{R}
VlnPlot(AllMB, features = c("SLC18A2"),pt.size = 0) +
        theme(axis.text = element_text(size=8),
              axis.title = element_text(size = 12),
              legend.text = element_text(size = 8),
              title = element_text(size = 12),
              legend.key.size = unit(0.4,"cm"))

SLC18A2Vln <- VlnPlot(AllMB, features = c("SLC18A2"),pt.size = 0) +
        theme(axis.text = element_text(size=8),
              axis.title = element_text(size = 12),
              legend.text = element_text(size = 8),
              title = element_text(size = 12),
              legend.key.size = unit(0.4,"cm"))

ggsave(SLC18A2Vln, filename = "Files/SLC18A2_Vln_scRNA_seq_Batch4.pdf", device = "pdf", width = 6, height = 4, units = "in")
```


These two neuronal subtypes could represent DA neurons from the substantia nigra and the ventral tegmental area. Therefore we will explore the expression of a set of literature derived markers across the two clusters.


CALB1 was historically associated with VTA DA neurons whereas GIRK2/KCNJ6,Sox6 was associated with SN DA neurons. According to a study in mice (see Defining Midbrain Dopaminergic Neuron Diversity by Single-Cell Gene Expression Profiling), SOX6 is indicative of DA1 type neurons (more SN-like). Neither ALDH1A1 or SLC17A6/Vglut2 are indicative in and of themselves of a neuron subtype, but co-expression of Vglut2 and SOX6 is indicative of a more SN-localised/SN-like DA neuron, and neurons expressed ALDH1A1 are reportedly more senstive in a mouse model of PD (see Classification of Midbrain Dopamine Neurons Using Single-Cell Gene Expression Profiling Approaches). 



```{r}
DANeuronSubtypesCellMarkerVlnPlot <- VlnPlot(AllMB, features = c("ALDH1A1","SOX6","CALB1","KCNJ6","SLC17A6"),pt.size = 0, ncol = 1) 

ggsave(DANeuronSubtypesCellMarkerVlnPlot,filename = "Files/DANeuronSubtypesCellMarkerVlnPlot.pdf", width = 20, height = 20)       
```



If one of these clusters corresponds to DA neurons/DA-like neurons then we might expect that the % of cells relative to the whole represented by this cluster would go down in PD patients relative to healthy controls. 

```{R}
table(Idents(AllMB), AllMB$case)
celltable <- table(Idents(AllMB), AllMB$case)
cell_type_number_df <- as.data.frame(celltable)
cell_type_number_df
```



```{R}
totalHC <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "HC"])
totalPD <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "PD"])
HCDAneuronsC1 <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "HC" & cell_type_number_df$Var1 == "Dopaminergic Neurons group 1"])
PDDAneuronsC1 <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "PD" & cell_type_number_df$Var1 == "Dopaminergic Neurons group 1"])
DAC1_Fishers_allcells <- cbind(totalHC - HCDAneuronsC1, totalPD - PDDAneuronsC1)
DAC1_Fishers_DAC1_Neurons <- cbind(HCDAneuronsC1, PDDAneuronsC1)

DAC1_Fishers_Table <- rbind(DAC1_Fishers_allcells, DAC1_Fishers_DAC1_Neurons)

colnames(DAC1_Fishers_Table) <- c("HC","PD")
rownames(DAC1_Fishers_Table) <- c("Total Non-DAC1","DAC1")

fisher.test(DAC1_Fishers_Table)

DAC1_Fishers_Table
```

```{R}
totalHC <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "HC"])
totalPD <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "PD"])
HCDAneuronsC2 <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "HC" & cell_type_number_df$Var1 == "Dopaminergic Neurons group 2"])
PDDAneuronsC2 <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "PD" & cell_type_number_df$Var1 == "Dopaminergic Neurons group 2"])
DAC2_Fishers_allcells <- cbind(totalHC - HCDAneuronsC2, totalPD - PDDAneuronsC2)
DAC2_Fishers_DAC2_Neurons <- cbind(HCDAneuronsC2, PDDAneuronsC2)

DAC2_Fishers_Table <- rbind(DAC2_Fishers_allcells, DAC2_Fishers_DAC2_Neurons)

colnames(DAC2_Fishers_Table) <- c("HC","PD")
rownames(DAC2_Fishers_Table) <- c("Total Non-DAC2","DAC2")

fisher.test(DAC2_Fishers_Table)

DAC2_Fishers_Table
```


```{R}
prop.table(table(Idents(AllMB), AllMB$case), margin = 2)
```


```{R}
HCOligo <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "HC" & cell_type_number_df$Var1 == "Oligodendrocytes"])
PDOligo <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "PD" & cell_type_number_df$Var1 == "Oligodendrocytes"])
Oligo_Fishers_allcells <- cbind(totalHC - HCOligo, totalPD - PDOligo)
Oligo_Fishers_Oligo_Neurons <- cbind(HCOligo, PDOligo)

Oligo_Fishers_Table <- rbind(Oligo_Fishers_allcells, Oligo_Fishers_Oligo_Neurons)

colnames(Oligo_Fishers_Table) <- c("HC","PD")
rownames(Oligo_Fishers_Table) <- c("Total Non-Oligodendrocytes","Oligodenrocytes")

fisher.test(Oligo_Fishers_Table)

Oligo_Fishers_Table
```

```{R}
HCAstro <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "HC" & cell_type_number_df$Var1 == "Astrocytes"])
PDAstro <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "PD" & cell_type_number_df$Var1 == "Astrocytes"])
Astro_Fishers_allcells <- cbind(totalHC - HCAstro, totalPD - PDAstro)
Astro_Fishers_Astro_Neurons <- cbind(HCAstro, PDAstro)

Astro_Fishers_Table <- rbind(Astro_Fishers_allcells, Astro_Fishers_Astro_Neurons)

colnames(Astro_Fishers_Table) <- c("HC","PD")
rownames(Astro_Fishers_Table) <- c("Total Non-Astrocytes","Astrocytes")

fisher.test(Astro_Fishers_Table)

Astro_Fishers_Table
```

```{R}
HCAstro <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "HC" & cell_type_number_df$Var1 == "Astrocytes"])
PDAstro <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "PD" & cell_type_number_df$Var1 == "Astrocytes"])
Astro_Fishers_allcells <- cbind(totalHC - HCAstro, totalPD - PDAstro)
Astro_Fishers_Astro_Neurons <- cbind(HCAstro, PDAstro)

Astro_Fishers_Table <- rbind(Astro_Fishers_allcells, Astro_Fishers_Astro_Neurons)

colnames(Astro_Fishers_Table) <- c("HC","PD")
rownames(Astro_Fishers_Table) <- c("Total Non-Astrocytes","Astrocytes")

fisher.test(Astro_Fishers_Table)

Astro_Fishers_Table
```



```{R}
HCOPCs <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "HC" & cell_type_number_df$Var1 == "OPCs"])
PDOPCs <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "PD" & cell_type_number_df$Var1 == "OPCs"])
OPCs_Fishers_allcells <- cbind(totalHC - HCOPCs, totalPD - PDOPCs)
OPCs_Fishers_OPCs_Neurons <- cbind(HCOPCs, PDOPCs)

OPCs_Fishers_Table <- rbind(OPCs_Fishers_allcells, OPCs_Fishers_OPCs_Neurons)

colnames(OPCs_Fishers_Table) <- c("HC","PD")
rownames(OPCs_Fishers_Table) <- c("Total Non-OPCs","OPCs")

fisher.test(OPCs_Fishers_Table)

OPCs_Fishers_Table

```


```{R}
HCGlu_GABA_Neurons <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "HC" & cell_type_number_df$Var1 == "Glu_GABA neurons"])
PDGlu_GABA_Neurons <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "PD" & cell_type_number_df$Var1 == "Glu_GABA neurons"])
Glu_GABA_Neurons_Fishers_allcells <- cbind(totalHC - HCGlu_GABA_Neurons, totalPD - PDGlu_GABA_Neurons)
Glu_GABA_Neurons_Fishers_Glu_GABA_Neurons_Neurons <- cbind(HCGlu_GABA_Neurons, PDGlu_GABA_Neurons)

Glu_GABA_Neurons_Fishers_Table <- rbind(Glu_GABA_Neurons_Fishers_allcells, Glu_GABA_Neurons_Fishers_Glu_GABA_Neurons_Neurons)

colnames(Glu_GABA_Neurons_Fishers_Table) <- c("HC","PD")
rownames(Glu_GABA_Neurons_Fishers_Table) <- c("Total Non-Glu_GABA neurons","Glu_GABA neurons")

fisher.test(Glu_GABA_Neurons_Fishers_Table)

Glu_GABA_Neurons_Fishers_Table
```


```{R}
HCMicroglia <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "HC" & cell_type_number_df$Var1 == "Microglia"])
PDMicroglia <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "PD" & cell_type_number_df$Var1 == "Microglia"])
Microglia_Fishers_allcells <- cbind(totalHC - HCMicroglia, totalPD - PDMicroglia)
Microglia_Fishers_Microglia_Neurons <- cbind(HCMicroglia, PDMicroglia)

Microglia_Fishers_Table <- rbind(Microglia_Fishers_allcells, Microglia_Fishers_Microglia_Neurons)

colnames(Microglia_Fishers_Table) <- c("HC","PD")
rownames(Microglia_Fishers_Table) <- c("Total Non-Microglia","Microglia")

fisher.test(Microglia_Fishers_Table)

Microglia_Fishers_Table
```


```{R}
HCEndothelialcells <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "HC" & cell_type_number_df$Var1 == "Endothelial cells"])
PDEndothelialcells <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "PD" & cell_type_number_df$Var1 == "Endothelial cells"])
Endothelialcells_Fishers_allcells <- cbind(totalHC - HCEndothelialcells, totalPD - PDEndothelialcells)
Endothelialcells_Fishers_Endothelialcells_Neurons <- cbind(HCEndothelialcells, PDEndothelialcells)

Endothelialcells_Fishers_Table <- rbind(Endothelialcells_Fishers_allcells, Endothelialcells_Fishers_Endothelialcells_Neurons)

colnames(Endothelialcells_Fishers_Table) <- c("HC","PD")
rownames(Endothelialcells_Fishers_Table) <- c("Total Non-Endothelial cells","Endothelial cells")

fisher.test(Endothelialcells_Fishers_Table)

Endothelialcells_Fishers_Table
```

```{R}
HCGlu_Neurons <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "HC" & cell_type_number_df$Var1 == "Glu_Neurons"])
PDGlu_Neurons <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "PD" & cell_type_number_df$Var1 == "Glu_Neurons"])
Glu_Neurons_Fishers_allcells <- cbind(totalHC - HCGlu_Neurons, totalPD - PDGlu_Neurons)
Glu_Neurons_Fishers_Glu_Neurons_Neurons <- cbind(HCGlu_Neurons, PDGlu_Neurons)

Glu_Neurons_Fishers_Table <- rbind(Glu_Neurons_Fishers_allcells, Glu_Neurons_Fishers_Glu_Neurons_Neurons)

colnames(Glu_Neurons_Fishers_Table) <- c("HC","PD")
rownames(Glu_Neurons_Fishers_Table) <- c("Total Non-Glu_Neurons","Glu_Neurons")

fisher.test(Glu_Neurons_Fishers_Table)

Glu_Neurons_Fishers_Table
```


```{R}
HCNaturalKillerCells <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "HC" & cell_type_number_df$Var1 == "Natural Killer Cells?? "])
PDNaturalKillerCells <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "PD" & cell_type_number_df$Var1 == "Natural Killer Cells?? "])
NaturalKillerCells_Fishers_allcells <- cbind(totalHC - HCNaturalKillerCells, totalPD - PDNaturalKillerCells)
NaturalKillerCells_Fishers_NaturalKillerCells_Neurons <- cbind(HCNaturalKillerCells, PDNaturalKillerCells)

NaturalKillerCells_Fishers_Table <- rbind(NaturalKillerCells_Fishers_allcells, NaturalKillerCells_Fishers_NaturalKillerCells_Neurons)

colnames(NaturalKillerCells_Fishers_Table) <- c("HC","PD")
rownames(NaturalKillerCells_Fishers_Table) <- c("Total Non-NaturalKillerCells","NaturalKillerCells")

fisher.test(NaturalKillerCells_Fishers_Table)

NaturalKillerCells_Fishers_Table
```


```{R}
HCnovel_Cluster14 <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "HC" & cell_type_number_df$Var1 == "novel_Cluster14"])
PDnovel_Cluster14 <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "PD" & cell_type_number_df$Var1 == "novel_Cluster14"])
novel_Cluster14_Fishers_allcells <- cbind(totalHC - HCnovel_Cluster14, totalPD - PDnovel_Cluster14)
novel_Cluster14_Fishers_novel_Cluster14_Neurons <- cbind(HCnovel_Cluster14, PDnovel_Cluster14)

novel_Cluster14_Fishers_Table <- rbind(novel_Cluster14_Fishers_allcells, novel_Cluster14_Fishers_novel_Cluster14_Neurons)

colnames(novel_Cluster14_Fishers_Table) <- c("HC","PD")
rownames(novel_Cluster14_Fishers_Table) <- c("Total Non-novel_Cluster14","novel_Cluster14")

fisher.test(novel_Cluster14_Fishers_Table)

novel_Cluster14_Fishers_Table
```

```{R}
HCnovel_Cluster16 <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "HC" & cell_type_number_df$Var1 == "novel_Cluster16"])
PDnovel_Cluster16 <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "PD" & cell_type_number_df$Var1 == "novel_Cluster16"])
novel_Cluster16_Fishers_allcells <- cbind(totalHC - HCnovel_Cluster16, totalPD - PDnovel_Cluster16)
novel_Cluster16_Fishers_novel_Cluster16_Neurons <- cbind(HCnovel_Cluster16, PDnovel_Cluster16)

novel_Cluster16_Fishers_Table <- rbind(novel_Cluster16_Fishers_allcells, novel_Cluster16_Fishers_novel_Cluster16_Neurons)

colnames(novel_Cluster16_Fishers_Table) <- c("HC","PD")
rownames(novel_Cluster16_Fishers_Table) <- c("Total Non-novel_Cluster16","novel_Cluster16")

fisher.test(novel_Cluster16_Fishers_Table)

novel_Cluster16_Fishers_Table
```

```{R}
HCGABA_Neurons <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "HC" & cell_type_number_df$Var1 == "GABA Neurons"])
PDGABA_Neurons <- sum(cell_type_number_df$Freq[cell_type_number_df$Var2 == "PD" & cell_type_number_df$Var1 == "GABA Neurons"])
GABA_Neurons_Fishers_allcells <- cbind(totalHC - HCGABA_Neurons, totalPD - PDGABA_Neurons)
GABA_Neurons_Fishers_GABA_Neurons_Neurons <- cbind(HCGABA_Neurons, PDGABA_Neurons)

GABA_Neurons_Fishers_Table <- rbind(GABA_Neurons_Fishers_allcells, GABA_Neurons_Fishers_GABA_Neurons_Neurons)

colnames(GABA_Neurons_Fishers_Table) <- c("HC","PD")
rownames(GABA_Neurons_Fishers_Table) <- c("Total Non-GABA Neurons","GABA Neurons")

fisher.test(GABA_Neurons_Fishers_Table)

GABA_Neurons_Fishers_Table
```

