---
title: "Batch1-8_Seurat_Harmony_Integration_and_Cluster_Assignment"
output: html_document
---

```{R}
library(Seurat)
library(cowplot)
library(ggplot2)
library(ggsci)
library(dplyr)
library(psych)
library(pheatmap)
library(harmony)
#library(clusterProfiler)
#library(org.Hs.eg.db)
library(DOSE)
library(GOSemSim)
library(enrichplot)
library(stringr)
```


Loading in the single cell data

```{r}
Batch1.data <- Read10X(data.dir = "~/Documents/Projects/PD5D_Repository/scRNA_and_scATAC-seq_Markdowns/batch1_ASAP_snRNA-Seq_021721/scRNA-seq/Files/cellranger_matrices/cellranger_aggr_matrices/filtered_feature_bc_matrix/")
Batch2.data <- Read10X(data.dir = "~/Documents/Projects/PD5D_Repository/scRNA_and_scATAC-seq_Markdowns/batch2_ASAP_snRNA-seq_031621/scRNA-seq/Files/cellranger_matrices/cellranger_aggr_matrices/filtered_feature_bc_matrix/")
Batch3.data <- Read10X(data.dir = "~/Documents/Projects/PD5D_Repository/scRNA_and_scATAC-seq_Markdowns/batch3_ASAP_snRNA-seq_031821/scRNA-seq/Files/cellranger_matrices/cellranger_aggr_matrices/filtered_feature_bc_matrix/")
Batch4.data <- Read10X(data.dir = "~/Documents/Projects/PD5D_Repository/scRNA_and_scATAC-seq_Markdowns/batch4_ASAP_snRNA-seq_040721/scRNA-seq/Files/cellranger_matrices/cellranger_aggr_matrices/filtered_feature_bc_matrix/")
Batch5.data <- Read10X(data.dir = "~/Documents/Projects/PD5D_Repository/scRNA_and_scATAC-seq_Markdowns/batch5_ASAP_snRNA-seq_042721/scRNA-seq/Files/cellranger_matrices/cellranger_aggr_matrices/filtered_feature_bc_matrix/")
Batch6.data <- Read10X(data.dir = "~/Documents/Projects/PD5D_Repository/scRNA_and_scATAC-seq_Markdowns/batch6_ASAP_snRNA-seq_043021/scRNA-seq/Files/cellranger_matrices/cellranger_aggr_matrices/filtered_feature_bc_matrix/")
Batch7.data <- Read10X(data.dir = "~/Documents/Projects/PD5D_Repository/scRNA_and_scATAC-seq_Markdowns/batch7_ASAP_snRNA-seq_050421/scRNA-seq/Files/cellranger_matrices/cellranger_aggr_matrices/filtered_feature_bc_matrix/")
Batch8.data <- Read10X(data.dir = "~/Documents/Projects/PD5D_Repository/scRNA_and_scATAC-seq_Markdowns/batch8_ASAP_snRNA-seq_050521/scRNA-seq/Files/cellranger_matrices/cellranger_aggr_matrices/filtered_feature_bc_matrix/")

```


```{R}
batch1samples <- read.csv(file.path("~/Documents/Projects/PD5D_Repository/scRNA_and_scATAC-seq_Markdowns/batch1_ASAP_snRNA-Seq_021721/scRNA-seq/Files/cellranger_matrices/cellranger_aggr_matrices", "aggregation.csv"), stringsAsFactors=F)
batch2samples <- read.csv(file.path("~/Documents/Projects/PD5D_Repository/scRNA_and_scATAC-seq_Markdowns/batch2_ASAP_snRNA-seq_031621/scRNA-seq/Files/cellranger_matrices/cellranger_aggr_matrices", "aggregation.csv"), stringsAsFactors=F)
batch3samples <- read.csv(file.path("~/Documents/Projects/PD5D_Repository/scRNA_and_scATAC-seq_Markdowns/batch3_ASAP_snRNA-seq_031821/scRNA-seq/Files/cellranger_matrices/cellranger_aggr_matrices", "aggregation.csv"), stringsAsFactors=F)
batch4samples <- read.csv(file.path("~/Documents/Projects/PD5D_Repository/scRNA_and_scATAC-seq_Markdowns/batch4_ASAP_snRNA-seq_040721/scRNA-seq/Files/cellranger_matrices/cellranger_aggr_matrices", "aggregation.csv"), stringsAsFactors=F)
batch5samples <- read.csv(file.path("~/Documents/Projects/PD5D_Repository/scRNA_and_scATAC-seq_Markdowns/batch5_ASAP_snRNA-seq_042721/scRNA-seq/Files/cellranger_matrices/cellranger_aggr_matrices", "aggregation.csv"), stringsAsFactors=F)
batch6samples <- read.csv(file.path("~/Documents/Projects/PD5D_Repository/scRNA_and_scATAC-seq_Markdowns/batch6_ASAP_snRNA-seq_043021/scRNA-seq/Files/cellranger_matrices/cellranger_aggr_matrices", "aggregation.csv"), stringsAsFactors=F)
batch7samples <- read.csv(file.path("~/Documents/Projects/PD5D_Repository/scRNA_and_scATAC-seq_Markdowns/batch7_ASAP_snRNA-seq_050421/scRNA-seq/Files/cellranger_matrices/cellranger_aggr_matrices", "aggregation.csv"), stringsAsFactors=F)
batch8samples <- read.csv(file.path("~/Documents/Projects/PD5D_Repository/scRNA_and_scATAC-seq_Markdowns/batch8_ASAP_snRNA-seq_050521/scRNA-seq/Files/cellranger_matrices/cellranger_aggr_matrices", "aggregation.csv"), stringsAsFactors=F)
```




```{R}
Batches <- ls()[grep("[[:digit:]].data",ls())]

createmetadata <- function(batch) {
  batchdf <- get(batch)
  tempcellcodes <- as.data.frame(batchdf@Dimnames[[2]])
  colnames(tempcellcodes) <- "barcodes"
  tempcellcodes$libcodes <- as.factor(gsub(pattern=".+-", replacement="", tempcellcodes$barcodes))
  samptable <- paste("batch",str_extract(batch,"[[:digit:]]"),"samples",sep="")
  tempcellcodes$samples <- as.vector(get(paste("batch",str_extract(batch,"[[:digit:]]"),"samples",sep=""))$library_id[tempcellcodes$libcodes])
  tempcellcodes$case <- as.vector(get(paste("batch",str_extract(batch,"[[:digit:]]"),"samples",sep=""))$case[tempcellcodes$libcodes])
  tempcellcodes$batch <- str_extract(batch,"Batch[[:digit:]]")
  return(tempcellcodes)
}

metadata <- bind_rows(lapply(Batches,createmetadata), .id = "column_label")

```


```{r}

Batch1to8_MTG <- CreateSeuratObject(counts = do.call(cbind,lapply(Batches,get)),
                            project = "Batch1to8_MTG",
                            min.cells = 3)
```


```{R}

Batch1to8_MTG@meta.data$sample_id <- metadata$samples
Batch1to8_MTG@meta.data$case <- metadata$case
Batch1to8_MTG@meta.data$batch <- metadata$batch

```


```{r}
Batch1to8_MTG[["percent.mt"]] <- PercentageFeatureSet(Batch1to8_MTG, pattern = "^MT-")

VlnPlot(Batch1to8_MTG, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size=0,split.by = "sample")
```



library(stats)
complete_metadata <- Batch1to8_MTG@meta.data
sample_nfeaturemeans 
#test <- group_by(complete_metadata,sample_id) %>% summarise(median(nFeature_RNA))


here <- group_by(complete_metadata,sample_id) %>% summarise(mad(nFeature_RNA, center = median(nFeature_RNA)))
mean(nfeature_RNA)
MAD <- mad(nfeature_RNA, center = median(nfeature_RNA))

MAD <- mad(nfeature_RNA, center = median(nfeature_RNA))
threeMAD <- (MAD*3)+median(Batch1to8_MTG@meta.data$nFeature_RNA)


average nfeature_RNA

```{R}
library(stats)
nfeature_RNA <- Batch1to8_MTG@meta.data$nFeature_RNA
mean(nfeature_RNA)
MAD <- mad(nfeature_RNA, center = median(nfeature_RNA))
threeMAD <- (MAD*3)+median(Batch1to8_MTG@meta.data$nFeature_RNA)

```


Filtering low quality cells:
```{r}
Batch1to8_MTG <- subset(Batch1to8_MTG, subset = nFeature_RNA > 200 & nfeature_RNA < threeMAD & percent.mt < 5)
```



Log Normalizing data: 
```{r}
Batch1to8_MTG <- NormalizeData(Batch1to8_MTG, normalization.method = "LogNormalize", scale.factor = 10000)
```

```{R}

Batch1to8_MTG <- FindVariableFeatures(Batch1to8_MTG, selection.method = "vst", nfeatures = 2000)

```

```{R}
saveRDS(Batch1to8_MTG,"Files/Batch1to8_MTG_preclustering")

```

center and scale data
```{r}
all.genes <- rownames(Batch1to8_MTG)
Batch1to8_MTG <- ScaleData(Batch1to8_MTG, features = all.genes, verbose = FALSE)
```


